### DISCUSSION TITLE
Is it advised to only use one of this or foxess_modbus, or is using both OK?

### DISCUSSION THREAD

**Discussion by azekeil** [nathanmarlor/foxess_em#288] [Q&A] [answered]
I have installed and set up https://github.com/nathanmarlor/foxess_modbus, (thank you!). I then found this and I'm wondering how it relates to the modbus integration. Does it use the modbus integration under the hood, or does it reimplement it, i.e. make additional calls over modbus and store the same data twice in your HA installation, and potentially cause race conditions in talking to the modbus slave?

Would be good to see an article highlighting the difference and offering guidance on which to use for what purpose. Thanks in advance!

**Comment by nathanmarlor** [comment]
They were designed to work together :)

**Reply by azekeil** [reply]
Does that mean I should install both, or that by installing foxess_em it will install foxess_modbus? What is actually going on under the hood? Can it lead to the race condition I mentioned in the question? Thanks :)

**Reply by nathanmarlor** [reply, marked as answer]
They're both self contained and accomplish different things independently. 

EM = control charging to optimise your off-peak periods.

Modbus = full inverter readings and control

No race conditions can occur, unless you're modifying charge periods from the modbus integration and undermining EM.

**Reply by jamesog** [reply]
While not a race condition, depending how frequently the modbus integration is polling there is a chance they can collide. Depending on what you're using to implement modbus (direct over serial, modbus TCP) sometimes things can get confused if two requests are made concurrently.

I'm using modbus TCP using https://github.com/3cky/mbusd and once in a while EM will try and make a change that must overlap with the modbus monitoring and it logs an error - but it looks like the EM integration retries and does the right thing 👍.

**Reply by azekeil** [reply]
Aah thanks, that's what I was wondering. 'Race condition' isn't quite the right word for it but modbus is a serial protocol with no innate collision management :)

**Reply by canton7** [reply]
Well, if you're using a TCP to Modbus converter, those will very often handle queuing requests from different sources so they don't collide. But sometimes they do a bad job and do things like return the wrong modbus response to the wrong requester.

**Reply by canton7** [reply]
@nathanmarlor I was wondering though, whether we could at least share a lock between EM and foxess_modbus, so they can't both make a request at the same time (actually sharing clients is probably a bit much). 

https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/singleton.py

**Reply by nathanmarlor** [reply]
@canton7 my first thought on reading this was using a shared lock in the hass object - it's singleton with a shared dictionary across all domains iirc. Looks like that's what the helper implements.

However I'm not sure it's an "us" problem - shouldn't this be a core part of the modbus device (W610, Waveshare etc)? Receiving mixed up responses would seem to point at an issue with the queuing logic (or lack of) within the device?

**Reply by canton7** [reply]
Oh, absolutely. We've put significant effort into working around the limitations of the adapters though

**Comment by koendruyts** [comment]
Dear I have also https://github.com/nathanmarlor/foxess_modbus V1.6.1 (latest) running with Modbus TCP connection and try to get the EM also working via Modbus TCP (same IP address and port settings to deafult) but EM seems not to accept this setting. No valid configuration. Maybe I do something wrong? Any advise welcome thx

**Reply by canton7** [reply]
What modbus adapter are you using? Can you screenshot the error message it shows you?

**Reply by koendruyts** [reply]
Dear, I have connected my LAN cable directly on the inverter in the additional port without any adapter. That's how it works for the foxess_modbus as well...see screenshot
![Schermafbeelding 2023-06-09 144632](https://github.com/nathanmarlor/foxess_em/assets/122438901/56e03a8d-e276-4e06-ae20-0aa2228bb59d)

**Reply by nathanmarlor** [reply]
@koendruyts EM won't work with a LAN connection as you can't modify the time period registers.
